### PR TITLE
.github/workflows: Introduce GitHub Actions Workflow integrations

### PR DESCRIPTION
**Pull request by nderjung**
### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.

### Base target

 - Architecture(s): N/A
 - Platform(s): N/A
 - Application(s): N/A

### Description of changes

This commit introduces GitHub-centric CI/CD definition that is intended to be executed on incoming PRs such that the incoming PR can be determined as working.  This supplementary file is based on-top of the [KraftKit GitHub Action Workflow Step][0] which is a native-to-Unikraft integration.  This integration allows for seamless building, packaging, distributing and executing (via TCG-based emulation) of Unikraft unikernels in a clear and declarative manner.

In this first integration pipeline, incoming PRs will be built against the following variant of the helloworld application:

- `qemu/x86_64`
- `qemu/arm64`
- `xen/x86_64`
- `fc/x86_64`
- `linuxu/x86_64`
- `linuxu/arm64`

Additionally, a more comprensive test will be executed with [libc-test][1] via uktest in emulation mode for qemu on `x86_64` only in order to display maximum coverage.

[0]: https://github.com/unikraft/kraftkit/blob/staging/action.yml
[1]: https://github.com/unikraft/lib-libc-test

Note: GitHub will not execute this action until it is merged into the target branch. A demonstration of this pipeline can be viewed here: https://github.com/nderjung/unikraft/actions/runs/5728684586 Therefore, to reproduce this work, simply apply this PR onto your fork on branch `staging` to see the action perform.

### CODE REVIEW

File: .github/workflows/integration.yaml [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+    - name: Fetch helloworld
+      uses: actions/checkout@v3
+      with:
+        repository: unikraft/app-helloworld
+        fetch-depth: 1
+        path: _helloworld
+
+    - name: Build and run libc-test via uktest
+      uses: unikraft/kraftkit@staging
```
Comment (craciunoiuc) [inline .github/workflows/integration.yaml:35]:
```suggestion
      uses: unikraft/kraftkit@v0
```

Why not like this and we do a tag?

File: .github/workflows/integration.yaml [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+    - name: Fetch helloworld
+      uses: actions/checkout@v3
+      with:
+        repository: unikraft/app-helloworld
+        fetch-depth: 1
+        path: _helloworld
+
+    - name: Build helloworld
+      uses: unikraft/kraftkit@staging
```
Comment (craciunoiuc) [inline .github/workflows/integration.yaml:172]: Same as above

File: .github/workflows/integration.yaml [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        submodules: recursive
+
+    - name: Fetch helloworld
+      uses: actions/checkout@v3
+      with:
+        repository: unikraft/app-helloworld
+        fetch-depth: 1
+        path: _helloworld
+
+    - name: Build helloworld
```
Comment (craciunoiuc) [inline .github/workflows/integration.yaml:171]: Why build both? Shouldn't this be run inside the app catalog by sending events somehow?

Does it mean that eventually we will also have all apps we want to check against, here?

**Review by craciunoiuc — COMMENTED** [review]
Looks good, some questions from my side

File: .github/workflows/integration.yaml (hunk 1 above)
Comment (nderjung) [reply .github/workflows/integration.yaml:35]: Because this represents a branch or tag, there is no branch `v0` or tag `v0` on the kraftkit repo.

File: .github/workflows/integration.yaml (hunk 3 above)
Comment (nderjung) [reply .github/workflows/integration.yaml:171]: Of course the catalog repo will run it as well. But this will be run once it is in already in `staging`. Here, we are checking the incoming PR builds against helloworld, on-demand.

Comment (craciunoiuc) [reply .github/workflows/integration.yaml:171]: Ok, so a double check, fine

File: .github/workflows/integration.yaml (hunk 1 above)
Comment (craciunoiuc) [reply .github/workflows/integration.yaml:35]: exactly, we could do it if you want ;)

A v0 branch wouldn't hurt, certainly easier to update than a tag

Should we do it?

Comment (nderjung) [reply .github/workflows/integration.yaml:35]: I fear a `v0` branch is extra maintenance on our side. Once we do a proper release in kraftkit, we can put `unikraft/kraftkit@stable`

**Comment by i-Pear** [comment]
I see there's "Unhandled Trap 13 (general protection), error code=0x0" in the demo, when running `lib-libc-test`. I also encountered this while testing my PR.

To solve this, you can try setting `LIBUKTEST_ALL` to `n` and `CONFIG_LIBUKTEST_TEST_MYSELF` to `n`. It seems the self-tests have some side effects, which can let `lib-libc-test` to crash.

File: .github/workflows/integration.yaml (hunk 1 above)
Comment (craciunoiuc) [reply .github/workflows/integration.yaml:35]: Okay 👍

**Comment by i-Pear** [comment]
For greater flexibility, each job can also be individually split into separate repositories/branches, making the `unikraft` repo more concise. Initially, `unikraft/test-workflows` was intended for this purpose, and you can be free to write into it as needed.

**Comment by i-Pear** [comment]
It may also be possible to save the test output into a text file and display the test results in a separate step. The extensive logs from `git clone` and `make` might make it difficult for users to locate the actual test results.

**Comment by nderjung** [comment]
Hi @i-Pear

> I see there's "Unhandled Trap 13 (general protection), error code=0x0" in the demo, when running `lib-libc-test`. I also encountered this while testing my PR.
>
> To solve this, you can try setting `LIBUKTEST_ALL` to `n` and `CONFIG_LIBUKTEST_TEST_MYSELF` to `n`. It seems the self-tests have some side effects, which can let `lib-libc-test` to crash.

Thanks for the suggestion. For this, I have force-pushed an update that now separates libc-test and "self-test" into two separate jobs. This can be viewed here:

https://github.com/nderjung/unikraft/actions/runs/5739682373

> For greater flexibility, each job can also be individually split into separate repositories/branches, making the `unikraft` repo more concise. Initially, `unikraft/test-workflows` was intended for this purpose, and you can be free to write into it as needed.

Separating jobs into different branches is definitely more overhead in terms of maintenance in my opinion. Instead this PR has introduced the use of a generic GitHub action tool that is also adoptable outside of testing Unikraft's core internally by anyone who simply wishes to build unikernels programmatically in a CI/CD context via GitHub Actions. Since KraftKit is the SDK behind this action, as we maintain this project we will always ensure that the action works [since we test it continuously internally](https://github.com/unikraft/kraftkit/actions/workflows/github-action.yaml) (see [here another example](https://github.com/unikraft/kraftkit/actions/runs/5727503369/job/15520734987#step:10:1)).

The `Kraftfile` has also been included in-line and represents a working expression of building Unikraft against helloworld. By keeping this declaration in-tree, modifications to Unikraft's core which affect this expression of KConfig options will be caught before they are made to `staging`. Additionally, new platforms, architectures, etc. introduced into the core must modify this file allowing the PR which introduces new functionality at this level to be tested within the scope of the same PR.

> It may also be possible to save the test output into a text file and display the test results in a separate step. The extensive logs from `git clone` and `make` might make it difficult for users to locate the actual test results.

One step at a time. We can parse the output later on as a quality-of-life improvement. Right now it's important just to get a build in and it is still viewable through GitHub's web UI. We can add additional wrappers ([see composites](https://docs.github.com/en/actions/creating-actions/creating-a-composite-action)) to make markdown summaries readable in the PR checks tab. I should note this is definitely planned, but we are waiting for [`ukdebug` to be refactored into two libraries, introducing a configurable `ukprint` library capable of outputting JSON-formatted logs](https://github.com/unikraft/unikraft/issues/665) which will allow for easier parsing of the results from `uktest`.

**Comment by craciunoiuc** [comment]
If needed, we could potentially limit the output to warnings, this should silence all the compilation messages

what do you think? @nderjung

**Comment by nderjung** [comment]
I would not remove the compilation steps from the CI at this stage as it is important to view what files are compiled and when warnings are emitted since [silencing warnings is an open issue and is often addressed in PRs](https://github.com/unikraft/unikraft/pulls?q=is%3Apr+silence+warning) and warning themselves need to be addressed individually.

I think the parsing of the compilation logs and execution logs should come later when we do these quality-of-life improvements. This can be addressed in the github-action itself and as PR in KraftKit.

**Review by craciunoiuc — APPROVED** [review]
Ok, let's get this in and let's see if you're lucky @nderjung 

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

**Review by razvand — APPROVED** [review]
Thanks, @i-Pear, for the comments. Let's merge this as it is now, and then we can improve it later.

Approved-by: Razvan Deaconescu <razvand@unikraft.io>
